docs/architecture/adr-062-p2p-architecture.md

```diff
@@ -0,0 +1,531 @@
+
+Primary design objectives have been:
+
+* Loose coupling between components, for a simpler, more robust, and test-friendly architecture.
+* Pluggable transports (not necessarily networked).
+* Better scheduling of messages, with improved prioritization, backpressure, and performance.
+* Centralized peer lifecycle and connection management.
+* Better peer address detection, advertisement, and exchange.
+* Backwards compatibility with the current P2P network protocols.
+
+The main abstractions in the new stack are:
+
+* `peer`: A node in the network, uniquely identified by a `PeerID` and stored in a `peerStore`.
+* `Transport`: An arbitrary mechanism to exchange raw bytes with a single peer using IO streams.
+* `Channel`: A bidirectional channel to asynchronously exchange Protobuf messages with peers addressed with `PeerID`.
+* `Router`: Maintains transport connections to relevant peers and routes channel messages.
+* Reactor: A design pattern loosely defined as "something which listens on a channel and reacts to messages".
+
+These abstractions and related concepts are illustrated in the following diagram and described in detail below.
+
+![P2P Architecture Diagram](img/adr-062-architecture.svg)
+
+### Transports
+
+Transports are arbitrary mechanisms for exchanging raw bytes with a peer. For example, a gRPC transport would connect to a peer over TCP/IP and send data using the gRPC protocol, while an in-memory transport might communicate with a peer running in another goroutine using internal byte buffers. Note that transports don't have a notion of a `peer` as such - instead, they communicate with an arbitrary endpoint address (e.g. IP address and port number), to decouple them from the rest of the P2P stack.
+
+Transports must satisfy the following requirements:
+
+* Be connection-oriented, and support both listening for inbound connections and making outbound connections using endpoint addresses.
+
+* Support multiple logical IO streams within a single connection, to take full advantage of protocols with native stream support. For example, QUIC supports multiple independent streams, while HTTP/2 and MConn multiplex logical streams onto a single TCP connection.
+
+* Provide the public key of the peer, and possibly encrypt or sign the traffic as appropriate. This should be compared with known data (e.g. the peer ID) to authenticate the peer and avoid man-in-the-middle attacks.
+
+The initial transport implementation will be a port of the current MConn protocol currently used by Tendermint, and should be backwards-compatible at the wire level. This will be followed by an in-memory transport for testing, and a QUIC transport that may eventually replace MConn.
+
+The `Transport` interface is:
+
+```go
+// Transport is an arbitrary mechanism for exchanging bytes with a peer.
+type Transport interface {
```

Comment:
Should we first describe what a `Connection` is?

Reply:
Yeah, I've been going back and forth on whether I should describe these concepts top-down or bottom-up. Bottom-up is more logical, but I feel like top-down may be easier to understand. No strong opinions though, happy to change if people feel it makes more sense.

Reply:
This is fine as-is.